packages/x-components/src/x-modules/scroll/components/main-scroll.vue

```diff
@@ -0,0 +1,403 @@
+   * the children position to be restored.
+   *
+   * Each child element that wants to have this support must be wrapped in a {@link MainScrollItem}
+   * component.
+   *
+   * @public
+   */
+  @Component({
+    mixins: [xComponentMixin(scrollXModule)]
+  })
+  /* eslint-disable @typescript-eslint/unbound-method */
+  export default class MainScroll extends NoElement {
+    /**
+     * The reference to the HTML node that has the scroll.
+     *
+     * @public
+     */
+    public $el!: HTMLElement;
+
+    /**
+     * If `true`, sets this scroll instance to the main of the application. Being the main
+     * scroll implies that features like restoring the scroll when the query changes, or storing
+     * the scroll position in the URL will be enabled for this container.
+     *
+     * @public
+     */
+    @Prop({ default: false, type: Boolean })
+    public useWindow!: boolean;
+
+    /**
+     * Timeout in milliseconds to abort trying to restore the scroll position to the target
+     * element.
+     *
+     * @public
+     */
+    @Prop({ default: 5000 })
+    public restoreScrollTimeoutMs!: number;
+
+    /**
+     * Intersection percentage to consider an element visible.
+     *
+     * @public
+     */
+    @Prop({ default: 0.5 })
+    public threshold!: number;
+
+    /**
+     * Adjusts the size of the scroll container bounds.
+     *
+     * @public
+     */
+    @Prop({ default: '' })
+    public margin!: string;
+
+    /**
+     * If true (default), sets the scroll position to top when an
+     * {@link XEventsTypes.UserAcceptedAQuery} event is emitted.
+     *
+     * @public
+     */
+    @Prop({
+      type: Boolean,
+      default: true
+    })
+    protected resetOnQueryChange!: boolean;
+
+    /**
+     * The elements that are currently considered visible.
+     *
+     * @internal
+     */
+    protected intersectingElements: HTMLElement[] = [];
+
+    /**
+     * Stores the identifier of the timeout that will consider the scroll failed to restore.
+     *
+     * @internal
+     */
+    protected restoreScrollFailTimeoutId?: number;
+
+    /**
+     * Pending identifier scroll position to restore. If it matches the {@link MainScrollItem.item}
+     * `id` property, this component should be scrolled into view.
+     *
+     * @internal
+     */
+    @State('scroll', 'pendingScrollTo')
+    public pendingScrollTo!: string;
+
+    /**
+     * Intersection observer to determine visibility of the elements.
+     *
+     * @returns An intersection observer to detect elements visibility.
+     * @internal
+     */
+    protected intersectionObserver: IntersectionObserver | null = null;
+
+    /**
+     * Creates an `IntersectionObserver` to detect the first visible elements. Children of this
+     * component should register themselves if they want to be observed.
+     *
+     * @returns The intersection observer.
+     * @public
+     */
+    @XProvide(FirstVisibleItemObserverKey)
```

Comment:
this key confused me twice already reviewing the PR. Wouldn't be clearer something like `MainScrollObserver` or just `ScrollObserver`?
we already know what is for